### ISSUE TITLE
Trying to figure out how to filter and reduce a relationship

### ISSUE THREAD

**Issue by GRardB** [JSONAPI-Resources/jsonapi-resources#830]
Hey!

First off, I'd like to say that this gem is remarkably good, and I can't express enough gratitude for everyone working on it. It's made my creating my personal project so much easier.

Now, I need some help! :grin:

Suppose I have the following relationship schema:

```
User                 Expenses
==================   ================
has_many :expenses   belongs_to :user
```

I'm wondering if it's possible to use JSONAPI to do this:

```
Get how much a user has spent (sum of expenses) during [e.g. Sept 2016]."
```

The idea is to calculate the sum of a portion of the user's expenses without exposing the expenses themselves to the client. I'm currently doing the calculation on the client-side, but because my app will show these numbers to other users, I want to make the expenses themselves private.

I already have a filter on the expense resource that allows me to get the expenses of a given month (for the user whom they belong to), but I can't figure out how to reduce the relationship into on value and make it property on the user resource.

In other words, I want to be able to do something like:

```
GET /users/:id/?filter[month]=:date
```

and get back:

```
{
.... All the user stuff,
monthlyExpenseTotal: 4893
}
```

I hope that makes sense. Any help would be really appreciated!!

**Comment by BanMido** [comment]
Hi Gerard,

I seems that you want some business/application logic applied on the data that you get from the JSONAPI. This is not the main thrust of JSONAPI.

As far as I understand, the intention of JSONAPI is to provide a standard way to access data/resources, whilst avoiding bike-shedding. You would typically use it as a raw data store, although you can augment that in a limited way using metadata. But the main mission of JSONAPI is to provide access to resources in a uniform way.

Is seems that you need another layer between your client and the JSONAPI server, which does the business/application logic calculations. That is how I see it.

You can choose to make the middle layer emit JSONAPI conformant JSON, if you so choose, with an intention of keeping the changes on client side minimal.

Just my 2 cents.

**Comment by GRardB** [comment]
Yeah, I guess that makes sense. I was actually chatting with a fellow Ember user on Slack, and he thought it would make more sense to have an endpoint like `/expense_reports/:user_id?month=:month` or something.

I have no issues writing a custom controller and skipping `jsonapi-resources` to do that, but I'm not sure how I'd format it to conform to the JSONAPI spec. Wherever I have custom logic, I've been doing stuff like:

```
render json: JSONAPI::ResourceSerializer.new(UserResource).serialize_to_hash(UserResource.new(User.find(id), context))
```

I'm not sure if this is the proper way to do it, but it does work. However, it requires a resource to exist for what I'm returning, and when I've tried creating an `ExpenseReport` resource, I get into all sort of issues since it doesn't correspond with an `ActiveRecord` model. I have tried creating a "model" using plain Ruby and including some `ActiveRecord` modules, but I still run into problems when the resource tries to call methods like `.all` and `.find_by_key`.

Is there a better way to format JSONAPI responses in custom controller methods?

**Comment by GRardB** [comment]
I think I've come up with a solution for what I need, at least for now. I'm simply going to compute the monthly expense totals and send them back in a hash along with the rest of the user. For example:

```
{
  id: 1,
  username: 'grardb',
  monthlyExpenseTotals: {
    <date>: 2000
  }
}
```

This will work for my needs, and at the very least will buy me some time to figure out something better if there's ever an issue of scale. First, people will need to use my app, though :grin:

Thanks!
